### PR TITLE
Add NadHAT PYBF405 board support

### PR DESCRIPTION
**Pull request by garatronic**
NadHAT PYBF405 is a very similar board than PYBV11 i did at the Raspberry Pi pHAT form factor. 
I propose to integrate my design into the native deposit. 
I just changed the almost obsolete NXP MMA7660 accelerometer for a Kionix KXTJ3. Add support for it.

Frederic Pierson
www.garatronic.fr

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution!

Please can you change the stm32f4xx_hal_conf.h file so it uses the new style; see `ports/stm32/boards/PYBV11/stm32f4xx_hal_conf.h` for an example/template.

Also, it'd be much better if the new accelerometer driver code reused as much of the existing accel.c code as possible.  Maybe all it needs is to adjust some constants and the way it reads the raw values?  Other methods should be able to stay the same.

**Comment by garatronic** [comment]
You're welcome.
I'll do my best in the future. For this one, the accelerometer is quite different, 16 bits data registers, no tilt register for smart detection. Hard to reuse the same code. I implemented to do the minimal common with MMA7660 and keep compatibility with PYBV11, at +/-2G setup and 8 bits mode, even if output value is different :
MMA7660 : 1.5g/32 = 0.047m/s 2
KXTJ3-1057 : 2g/128 = 0.016 m/s 2

File: ports/stm32/accel.c [hunk 1]
```diff
@@ -226,3 +232,180 @@ const mp_obj_type_t pyb_accel_type = {
...
+#define I2C_TIMEOUT_MS (50)
+
+#define KXT_ADDR            (0x0F)
+#define KXT_REG_DCST_RESP   (0x0C)
+#define KXT_REG_WHO_AM_I    (0x0F)
+#define KXT_REG_XOUT_H      (0x07)
+#define KXT_REG_YOUT_H      (0x09)
+#define KXT_REG_ZOUT_H      (0x0B)
+#define KXT_REG_CTRL_REG1   (0x1B)
+#define KXT_REG_CTRL_REG2   (0x1D)
```
Comment (dpgeorge) [inline ports/stm32/accel.c]: If you rename `KXT_ADDR` -> `MMA_ADDR`, `KXT_REG_XOUT_H` -> `MMA_REG_X`, etc, then you can reuse a lot of the existing functions (like `read_axis`, `pyb_accel_x` etc).

File: ports/stm32/accel.c [hunk 2]
```diff
@@ -226,3 +232,180 @@ const mp_obj_type_t pyb_accel_type = {
...
+STATIC mp_obj_t pyb_accel_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    // check arguments
+    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+
+    // init accel object
+    pyb_accel_obj.base.type = &pyb_accel_type;
+    accel_start();
+
+    return MP_OBJ_FROM_PTR(&pyb_accel_obj);
+}
```
Comment (dpgeorge) [inline ports/stm32/accel.c]: This function is the same as the existing one so doesn't need to be duplicated.

File: ports/stm32/accel.c [hunk 3]
```diff
@@ -226,3 +232,180 @@ const mp_obj_type_t pyb_accel_type = {
...
+
+/// \method tilt()
+/// Not implemented with KXTJ3 accelerometer
+
+/// \method filtered_xyz()
+/// Not implemented with KXTJ3 accelerometer
+
+/// \method xyz()
+/// Get a 3-tuple of x, y and z values.
+STATIC mp_obj_t pyb_accel_xyz(mp_obj_t self_in) {
```
Comment (dpgeorge) [inline ports/stm32/accel.c]: I would call this just `filtered_xyz`.  It doesn't matter if it's not filtered, at least it gives the same API as the MMA.

**Review by dpgeorge — COMMENTED** [review]
It is possible to eliminate mote of the code duplication in accel.c.  For things that are different they can be individually selected throughout the file via `#if MICROPY_HW_HAS_KXTJ3`.

File: ports/stm32/accel.c (hunk 3 above)
Comment (garatronic) [reply ports/stm32/accel.c]: Done, i did changes and keep some filtering with the .filtered_xyz()
Thanks for your advice.

File: ports/stm32/accel.c [hunk 4]
```diff
@@ -185,22 +259,60 @@ STATIC mp_obj_t pyb_accel_filtered_xyz(mp_obj_t self_in) {
...
+            data_sum[i] += (int8_t)data[i*2];
+        }
+        // no need to waste time for the last loop 
+        if (j >= FILT_DEPTH - 1) {
+            break;
+        }
+        // ODR is 200Hz with 10ms LP filter roll-over
+        mp_hal_delay_ms(10+1);
+        j++;
+    }
```
Comment (dpgeorge) [inline ports/stm32/accel.c]: This filtering routine accumulates all the FILT_DEPTH samples each time the method is called.  But for the MMA it just takes one sample per call and stores the results in the rolling buffer "buf".  Regardless of whether one approach is better or not, it should be that the KXT uses the same algorithm as the MMA.  It should be simple to do this, just reuse the MMA code and select with a macro how the data is read in for the filtering.

Comment (garatronic) [reply ports/stm32/accel.c]: Yep, you're right. I did the changes in the code. Only one access to the component per method call.

**Comment by dpgeorge** [comment]
Thanks for updating the filter code.

I did some more cleanup and posted a version of this PR at #5217, which makes the accelerometer driver a bit more generic, so needs less changes to add support for the KXTJ3.

**Comment by dpgeorge** [comment]
This was squashed and merged in 3105207afff2ec0b22ecbdea756be5bc5006daff through 925f244ab39d52901b402695a9a8b0bfce6415f7, as part of #5217

**Pull request closed without merging**